samples/HttpClientFactorySample/Program.cs

```diff
@@ -53,6 +54,19 @@ public static void Configure(IServiceCollection services)
                 c.DefaultRequestHeaders.Add("Accept", "application/vnd.github.v3+json"); // Github API versioning
                 c.DefaultRequestHeaders.Add("User-Agent", "HttpClientFactory-Sample"); // Github requires a user-agent
             })
+
+            // Build a totally custom policy using any criteria
+            .AddPolicyHandler(Policy.Handle<HttpRequestException>().RetryAsync())
+
+            // Build a policy that will handle exceptions (connection failures)
+            .AddExceptionPolicyHandler(p => p.RetryAsync())
+
+            // Build a policy that will handle exceptions and 500s from the remote server
+            .AddServerErrorPolicyHandler(p => p.RetryAsync())
+
+            // Build a policy that will handle exceptions, 400s, and 500s from the remote server
+            .AddBadRequestPolicyHandler(p => p.RetryAsync())
```

Comment:
This is probably the best place to start the discussion. 

I understand the requirements for building a message handler to be:
- what kinds of things do I consider an error?
- what do I want to do when I find one?

I think there's some value in us providing guidance about the first question (what should I consider an error), and I think the second question is easy enough for users to answer (`p => p.RetryAsync()`).

I've defined a few flavors of extension methods here. I think `AddServerErrorPolicyHandler` is probably going to be the most useful balance between "sane default" and hiding complexity. Most folks will always want to treat a connection failure or a 500 as an error.

If you use `AddPolicyHandler` to build `AddServerErrorPolicyHandler` it will end up looking like:
```
.AddPolicyHandler(
  Policy
    .Handle<HttpRequestException>()
    .OrResult<HttpResponseMessage>(r => r.StatusCode >= HttpStatusCode.InternalServerError)
    .RetryAsync());
```
I think it's a good enough for us to find an attractive name for this and include it 👍 

A few questions for discussion:
- Handlers may be long-lived or short-lived. Is it important we reuse policies? It is important that we don't?
- There's a little bit of tension around whether or not things should be `IAsyncPolicy` or `IAsyncPolicy<T>` in some of the simple cases, so I planned to support both. Is this a good plan?

Reply:
I jump in even if not pinged. Policies' lifecycle should be independent from the handler's one because there are policies like the short-circuit one that are designed to be "singleton-ish".

Ideally policies should be attached to freshly generated handler's but it feels like it leads to a pool manager for the policies to be used by a pool manager for the handlers...

Reply:
@Kralizek thanks!

The factory already provides some pooling and lifecycle management for handlers (docs in progress). 

- It's easy to make a policy a singleton. 
- It's also easy to make a policy have the same lifetime as a handler. 
- It would be a bit tricky to do something different, and I'd like to avoid it if possible

> Policies' lifecycle should be independent from the handler's one because there are policies like the short-circuit one that are designed to be "singleton-ish".

This makes intuitive sense to me. 

Are there real use cases for making policies that have the same lifetime as the handler? Should they all be singleton?

Reply:
Responding to @rynowak and @Kralizek on policy lifecycle:

**TL;DR** Policies should be single-instance for each policy configured for use with a given named `HttpClient` configuration.  

For example, if a Polly policy for a given named `HttpClient` is config'd by supplying a `createPolicy()` factory Func, that `createPolicy()` func should be used to create one instance of that policy, and that single policy instance supplied to all subsequently created handlers/`HttpClient` instances of that named configuration. 

---

#### Explanation

Two policy types maintain state across calls for their designed function: **circuit-breakers** (tracking call faults/successes) and **bulkheads** (limiting parallelism).  

If `HttpClientFactory` were to create fresh instances when recycling handlers, that state would suffer either discontinuity (if new instances replaced old instances 'instantaneously') or fragmentation/multiplicity (if new and old instances co-existed).

eg **Bulkhead:** Generating fresh BulkheadPolicy instances for new handlers could cause executions to exceed the parallelism configured by the developer, as double/multiple capacity would be available while both newly-spun-up and existing / being-retired instances co-existed.  

No other policies maintain state, but all Polly policies are entirely thread-safe and memory-efficient for long lifetime, so there's no real case for shorter lifetimes; it'd just be more allocations and GCs.